zeppelin-server/src/test/java/org/apache/zeppelin/service/InterpreterServiceTest.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zeppelin.service;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.when;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import org.apache.commons.io.FileUtils;
+import org.apache.zeppelin.conf.ZeppelinConfiguration;
+import org.apache.zeppelin.dep.DependencyResolver;
+import org.apache.zeppelin.interpreter.InterpreterSettingManager;
+import org.apache.zeppelin.notebook.socket.Message;
+import org.apache.zeppelin.rest.message.InterpreterInstallationRequest;
+import org.apache.zeppelin.socket.NotebookServer;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class InterpreterServiceTest {
+  @Mock ZeppelinConfiguration mockZeppelinConfiguration;
+  @Mock NotebookServer mockNotebookServer;
+  @Mock InterpreterSettingManager mockInterpreterSettingManager;
+
+  Path temporaryDir;
+  Path interpreterDir;
+  Path localRepoDir;
+
+  InterpreterService interpreterService;
+
+  @Before
+  public void setUp() throws Exception {
+    temporaryDir = Files.createTempDirectory("tmp");
+    interpreterDir = Files.createTempDirectory(temporaryDir, "interpreter");
+    localRepoDir = Files.createTempDirectory(temporaryDir, "local-repo");
+
+    when(mockZeppelinConfiguration.getInterpreterDir()).thenReturn(interpreterDir.toString());
```

Comment:
We can set java properties via System.setProperty instead of usign mock ZeppelinConfiguration here

Reply:
Yes, right. But in case of `InterpreterService`, ZeppelinConfiguration is passed as a parameter of a constructor, - personally, I think a static method is hard to test it -, thus I used mocking object of ZeppelinConfiguration.

Reply:
Not sure what you mean. I don't think here we need static method. Just call System.setProperty in setUp should work. Or do I miss anything ?

Reply:
Yes, Yes, What I mean is in most cases, ZeppelinConfiguration is created inside any class by `ZeppelinConfiguration.create()`. But I don't think it's a good pattern to test it, so I added ZeppelinConfiguration as a parameter and injected it when creating `InterpreterService`.

Reply:
Not sure whether I understand you. Here you can just do it as following which looks more easier and don't need mock. 
```
System.setProperty(ConfVars.ZEPPELIN_INTERPRETER_DIR.getVarName(), interpreterDir.toString());
System.setProperty(ConfVars.ZEPPELIN_INTERPRETER_LOCALREPO.getVarName(), localRepoDir.toString());
ZeppelinConfiguration zConf = new ZeppelinConfiguration();
interpreterService =
       new InterpreterService(zConf, mockInterpreterSettingManager);
```

Reply:
Yes, right. We can do it. Regardless of static method, as I think, it's more fragile to use `System.setProperty()` because it might be a problem if we missed to clear or restore the value. Isn't it?

Reply:
hmm, right. Never mind, this is not a critical issue, I am fine with your current approach.

Reply:
Yes, it's not a big problem. You and I might have talked about some different approaches to seeing the code.

Reply:
I feel a bit sorry because I should have explained it more details at the first time. I assumed you might know my thoughts a bit.